core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala

```diff
@@ -0,0 +1,235 @@
+import scala.io.{Codec, Source}
+
+import com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileStatus, FileSystem, Path}
+import org.json4s.jackson.JsonMethods.parse
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN
+import org.apache.spark.scheduler._
+import org.apache.spark.util.{JsonProtocol, Utils}
+
+class EventLogFileCompactor(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    fs: FileSystem) extends Logging {
+
+  private val maxFilesToRetain: Int = sparkConf.get(EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN)
+
+  // FIXME: javadoc - caller should provide event log files (either compacted or original)
+  //  sequentially if the last event log file is already a compacted file, everything
+  //  will be skipped
+  def compact(eventLogFiles: Seq[FileStatus]): Seq[FileStatus] = {
+    if (eventLogFiles.isEmpty) {
+      return Seq.empty[FileStatus]
+    }
+
+    // skip everything if the last file is already a compacted file
+    if (EventLogFileWriter.isCompacted(eventLogFiles.last.getPath)) {
+      return Seq(eventLogFiles.last)
+    }
+
+    val (filesToCompact, filesToRetain) = findFilesToCompact(eventLogFiles)
+    if (filesToCompact.isEmpty) {
+      filesToRetain
+    } else {
+      // first pass
+      val bus = new ReplayListenerBus()
+
+      val builders = ServiceLoader.load(classOf[EventFilterBuilder],
+        Utils.getContextOrSparkClassLoader).asScala
+      builders.foreach(bus.addListener)
+
+      filesToCompact.foreach { log =>
+        Utils.tryWithResource(EventLogFileReader.openEventLog(log.getPath, fs)) { in =>
+          bus.replay(in, log.getPath.getName)
+        }
+      }
+
+      // second pass
+      val rewriter = new FilteredEventLogFileRewriter(sparkConf, hadoopConf,
+        filesToCompact, fs, builders.map(_.createFilter()).toSeq)
+      rewriter.start()
+      rewriter.rewrite()
+      rewriter.stop()
+
+      // cleanup files which are replaced with new compacted file.
+      cleanupCompactedFiles(filesToCompact)
+
+      fs.getFileStatus(new Path(rewriter.logPath)) :: filesToRetain.toList
+    }
+  }
+
+  private def cleanupCompactedFiles(files: Seq[FileStatus]): Unit = {
+    files.foreach { file =>
+      try {
+        fs.delete(file.getPath, true)
+      } catch {
+        case _: IOException => logWarning(s"Failed to remove ${file.getPath} / skip removing.")
+      }
+    }
+  }
+
+  private def findFilesToCompact(
+      eventLogFiles: Seq[FileStatus]): (Seq[FileStatus], Seq[FileStatus]) = {
+    val lastCompactedFileIdx = eventLogFiles.lastIndexWhere { fs =>
+      EventLogFileWriter.isCompacted(fs.getPath)
+    }
+    val files = eventLogFiles.drop(lastCompactedFileIdx)
+
+    if (files.length > maxFilesToRetain) {
+      (files.dropRight(maxFilesToRetain), files.takeRight(maxFilesToRetain))
```

Comment:
This will compact all the files at once under `maxFilesToRetain`. Do we need to be so aggressive? Is it not possible to split the compaction and let other functionalities to run.